### DISCUSSION TITLE
Nodejs event emitter implementation in Nextjs 13 or 14

### DISCUSSION THREAD

**Discussion by torver213** [vercel/next.js#60534] [App Router]
How do i make nodejs event emitter object to listen to emitted events globally within my nextjs13 application just like in normal nodejs environment?

I have a file *post.js* where a user creates a new post. I want to promptly return the response back to the user, emit an event where a handler function will listen to the emitted event in another file & perform some other tasks in the background like updating post tags count and then finally  emitting an sse(server sent event) to client app to notify other connected clients.

First of all, i created a file known as *eventEmitter.js*  where I created a single instance to be used within my application below;

```ts
    //eventEmitter.js

    import { EventEmitter } from 'node:events';

    const eventEmitter = new EventEmitter({captureRejections: true})

    export default eventEmitter;

```

```ts

    //post.js
  
    import eventEmitter from './eventEmitter'
  
    const insertPost = async() => {
       // perform db post insert
       
       //emit an event to be listened to perform a background task
       eventEmitter.emit("PING", "hello world");
       
       //return 
       return {id: 1, title: 'hello' }
   }
   

   insertPost()
   
   
    // this only works when the listener is in the same file with the emitted event
   eventEmitter.on("PING", (ev) => {
        console.log(ev)
   });
   
```

```ts

    //listener.js
  
    import eventEmitter from './eventEmitter'
    // this doesn't work i.e the event is not captured in the file
   eventEmitter.on("PING", (ev) => {
        console.log(ev)
   });
```

```ts
 //route.js
  export function GET(request: Request) {
       const responseStream = new TransformStream();
       const writer = responseStream.writable.getWriter();
       const encoder = new TextEncoder();
       //... other code
      
       // this doesn't work i.e the event is not captured in the file
      eventEmitter.on("PING", (ev) => {
        console.log(ev)
     });
     
     // my streaming logic
}
```

My question is, how do i listen to the emitted event in other files like **listener.js** and **route.js** files  other than just **post.js** file?

I want to listen to the emitted events globally within my nextjs app 

The above works seamlessly in a nodejs environment and i expected it to work in nextjs environment too.

*PS:* I'm using nextjs App Router

Please help.

**Comment by jonshipman** [comment]
I mean, couldn't you have a src/lib/application.ts and listen from there?

```
import { EventEmitter } from 'node:events';

class ApplicationEmitter extends EventEmitter {}

export const app = new ApplicationEmitter();

app.on('something', function() {
	// do something...
});
```

The issue you're running into is that on that GET request, it only fires during the polling. Not knowing where the app is hosted is another issue as "serverless" hosts spin up instances on request. If you are on a traditional server then you can just import '@/lib/application' into 'src/middleware.ts' and it should work with emits.

Also, it's never a good idea to wrap listeners in functions. That's a quick way to a memory leak my friend. Instead emit a callback. e.g. `eventEmitter.emit('PING', function callback() { /* awesome */ });` then in the application.ts, `eventEmitter.on('PING', function(cb) { cb(); });`

**Comment by latifs** [comment]
Hi @techfortified, 
have you been able to solve that?
thanks

**Comment by arvilmena** [comment]
I tried putting it on `instrumentation.ts` but I am getting errors for when webpack is trying to bundle server packages. Even though I have these packages ignore in the nextjs webpack config. Is bundling intrumentation.ts skips the webpack config?

Curious as how others are able to implement eventEmitters in NextJs server

**Reply by jonshipman** [reply]
You need to inline the imports. `const cron = await import('node-cron');` within the register function.

```
export async function register() {
  if (process.env.NEXT_RUNTIME === 'nodejs') {
    await import('@/server');
```

**Reply by arvilmena** [reply]
@jonshipman Thanks, that led me to fixing the build error in my actual project, but in the end, the events are still not getting listened.

So, I created a minimal repro repo: https://github.com/arvilmena/test--nextjs--eventemitter

There's a server action that will emit an event. And some event listener inside my app should be invoked everytime that event was emitted. I tried 2 approach:

1. Is to instantiate the event listeners in the `instrumentation.ts` = did not work
2. Is to instantiate the event listeners at the top of the server action file = it worked! but looks ugly, it looks like that the event listeners will be instantiated everytime any server action in that server action file is called? will that cause memory leak?

If you can take a look here's the reproducible repo: https://github.com/arvilmena/test--nextjs--eventemitter
Branch for attempt 1: https://github.com/arvilmena/test--nextjs--eventemitter/tree/attempt/1-via-instrumentation-js 
Branch for attempt 2: https://github.com/arvilmena/test--nextjs--eventemitter/tree/attempt/2-via-on-top-of-server-actions-file

Can you guide me on what's the best approach?

Thank you!

**Reply by jonshipman** [reply]
You should be doing the instrumentation like this:

```
export async function register() {
  if (process.env.NEXT_RUNTIME === 'nodejs') {
    console.log(`Hello from instrumentation.ts!`);
    const { initEventEmitter } = await import('./server-events/initEventEmitter');
    initEventEmitter();
  }
}
```

However, your understanding of Next.js is flawed. You're not running a single instance of the application. When you click "Get me a random pokemon" the event that fires is not on the same process as the instrumentation.js. 

You need a shared space for reading and writing data. The most common is through a database of some sort. MongoDB (for example) has a realtime socket layer.

For the approach you have implemented, I think the best approach would be to fire something in the `getRandomPokemon` function here: `src\actions\getRandomPokemon.ts`. So take the event emitter out of the equation and have the `// TODO: log to the database` as part of the function.


```
export async function Pokemon() {
  const log = async function(...args) {
    // TODO: Log to the database
    
    console.log(Date.now(), ...args);
  }

  return {
    log,
    async random() {
      const name = pokemon.random();
      
      await log(name);
      return name;
    }
  }
}
```

**Reply by arvilmena** [reply]
@jonshipman thanks a bunch for the insight. With that approach, it looks like the side-effects, `log()` in your example, are now blocking the early return of the function. And that the side effect error handling codes will dilute the function, meanwhile in event emitter-listener pattern, those codes are somewhere else, and errors thrown doesn't affect the return value of the function.

But what you've confirmed is that the server action process is different, if this is isolated from the main app, is it fine to just put my listeners on top of the server action file? Considering that process would be garbage cleaned up after all the side-effects and event listeners have finished?

I wonder how can I load test this to confirm there's no memory leak. Is the instance where the server actions are running can be obtained via `ps`?

**Reply by jonshipman** [reply]
Generally no, you'll be creating new listeners frequently. Truth be told, if you are inflexible in as far as awaiting the log on the server then your best option would be to deploy an express server for your REST actions. Then you will have complete control over the lifecycle of your application. You do have the option of skipping the await in the random() function, but it would be less stable. In the end, you're arguing over AT MOST a 1ms delay in response - most likely less - for a simple db write. 

Next.js (and the server actions) work best when treated imperatively.
